### DISCUSSION TITLE
Using a copy of pyinstaller managed by pipx

### DISCUSSION THREAD

**Discussion by pfmoore** [pyinstaller/pyinstaller#7955] [Help] [answered]
Generally, I prefer *not* to install my development tools in my application virtual environment, but rather install them globally using pipx. So, I set up a copy of pyinstaller that way.

However, when I try to build a simple application, pyinstaller can't find my dependencies. To be specific, I have a project directory `test`, containing `test.py`, which looks like this:

```
import rich

rich.print("[red]Hello, [cyan]world!")
```

I also have a virtual environment `.venv` in the project directory, which has `rich` installed into it. `.venv\Scripts\python test.py` works fine.

When I run `pyinstaller test.py`, this creates a `test.exe` file in the `dist` directory - but when I run it, I get the error:

```
Traceback (most recent call last):
  File "test.py", line 1, in <module>
    import rich
ModuleNotFoundError: No module named 'rich'
[46664] Failed to execute script 'test' due to unhandled exception!
```

First of all, I'm surprised that pyinstaller didn't report that rich wasn't available when I built the application. But more importantly, how do I get pyinstaller to know that my application's dependencies are available in the virtual environment?

I couldn't find much information in the pyinstaller documentation on how to define script dependencies, which might mean that I've missed something important - or it might mean I'm thinking about the problem in the wrong way. I have a suspicion that I could do something like this by modifying the `.spec` file, but (a) I couldn't find much documentation about what the arguments to the `Analysis` class are (I'm wondering if `pathex` is the one to use here) and (b) this seems like a rather "advanced" thing to be doing for what feels to me like a fairly basic usage.

Is having pyinstaller installed separately simply not a supported approach? That seems somewhat annoying, as I strongly prefer to keep runtime and development-time dependencies independent. Would I have similar problems if I tried to run pyinstaller from a nox (or tox) session?

**Comment by rokm** [comment, marked as answer]
> Is having pyinstaller installed separately simply not a supported approach?

It is indeed not supported, because PyInstaller is collecting from the environment in which it is installed. 

And if you are in habit of using both global and virtual environment and end up with PyInstaller installed in both, you need to take care to invoke the correct `pyinstaller` shim (or better yet, use `python -m PyInstaller ...` to ensure you are using the version in the currently active environment).

**Reply by pfmoore** [reply]
That's a shame. So there's no way to tell PyInstaller to collect from a different environment? I really dislike having to keep multiple versions of my development tools up to date. I routinely come back to projects after months of inactivity, and I *never* think to update the tools in the environment. That's basically why I started using a "globally installed tools" approach.

**Reply by rokm** [reply]
If you really want to go down that route, you could adjust the search paths - either using [`--paths`](https://pyinstaller.org/en/stable/usage.html#cmdoption-p) or (preferably) just modifying `PYTHONPATH` before running PyInstaller. 

We tend to discourage users from doing that because it is usually the sign that they are unknowingly mixing their (virtual) environments, but it looks like that is exactly what you want. Just keep in mind that the virtual environment needs to be of the same python (minor) version as the PyInstaller's environment, otherwise you will run into issues with extensions...

**Reply by pfmoore** [reply]
One reason I want to do this is actually to be sure I only have a "clean" environment. For example, I tried to build a standalone copy of pip (not something we'd actually need, but it was as a proof of concept of a non-trivial app). The built version contained the pyinstaller `.dist-info` file in the `_internal` directory, which seems very wrong.

I think ultimately, there's a fundamental difference in viewpoints here. I don't have a need for the sort of space optimisation that scanning imports provides, so I'd much rather just say "here's a set of requirements from PyPI that my app needs, please bundle them into an executable". Given that PyInstaller works by scanning imports, the compromise I was hoping for is to do this by building a virtual environment with just those requirements, then building the app from that environment. That gives me the control that I want over what PyInstaller sees, although it still leaves me needing to fix up any imports that get missed because of things like `importlib.import_module` (but thankfully, that's rare). But it seems that's not going to work either.

As I say, it's not a real-life issue at this point - I'm really just trying out PyInstaller to see if I can make it fit into the sort of workflow I prefer. The answer seems to me "maybe, but it's a lot of work fighting the mismatch between the tool's expectations and mine".

Thanks for your time, though. The information has been useful to me.

**Reply by bwoodsend** [reply]
>  tried to build a standalone copy of pip (not something we'd actually need, but it was as a proof of concept of a non-trivial app)

`subprocess.run([sys.executable, ...])` (which I recall pip using a fair bit) is a bit of a disaster in PyInstaller because `sys.executable` is your application rather than `python.exe`. In retrospect, I wish we set `sys.executable` to `None` and stored the executable's path in a different variable but that's a breaking change that I'm not sure I can justify making.

> The built version contained the pyinstaller .dist-info file in the _internal directory, which seems very wrong.

That is very wrong. How did you get there? Just running `pyinstaller pip/__main__.py`?

> If you really want to go down that route, you could adjust the search paths - either using [--paths](https://pyinstaller.org/en/stable/usage.html#cmdoption-p) or (preferably) just modifying PYTHONPATH before running PyInstaller.

This will mean that your original Python environment is also in PyInstaller's search paths which rather defeats the purpose of using the clean virtual environment. Installing PyInstaller into the virtual environment is the preferred approach. In all honesty, given how much of a mess our users get in when they try to configure search paths themselves (it's amazing how many people don't understand the `sys.path` search process), I wouldn't want to do anything to endorse it.

**Reply by pfmoore** [reply]
> That is very wrong. How did you get there? Just running pyinstaller pip/__main__.py?

I believe so, but I'd been messing around trying things out, so I wouldn't waste too much time worrying about it. If it ever becomes a real concern, I'll raise a proper bug report.

And yes, I 100% agree that making a standalone exe out of pip is going to be far more trouble than just throwing PyInstaller at it (we dynamically import all of the command modules, we install into the environment pip is installed in by default, and like you say we rely heavily on being able to run scripts with the environment's Python executable). As I say, it was just the quickest example of a "more complicated app" that I had to hand.

> Installing PyInstaller into the virtual environment is the preferred approach.

Understood. Thanks again for the helpful discussion.

**Reply by rokm** [reply]
> > The built version contained the pyinstaller .dist-info file in the _internal directory, which seems very wrong.
> 
> That is very wrong. How did you get there? Just running `pyinstaller pip/__main__.py`?

Hmm, this doesn't seem to happen on linux, but it does on Windows. 

Specifically, the offending import chain looks as follows: `pip._vendor.pkg_resources` -> `__main__` -> `PyInstaller.__main__` -> everything PyInstaller. I think I recall seeing a guard against pulling in `__main__` somewhere; will dig around a bit more to see what is going on... 

I guess this supports your assessment that import scanning can be brittle, though.
